[Weather Underground] Error invoking #valueOf(String)

OH 2.3.0 snapshot 1231
WU binding 0.10.0.201803161254

I don't know if this is an issue with the binding, [Units of Measurement](https://github.com/eclipse/smarthome/issues/5243), or my setup. All of my other weather items using UOM are working, but two of my items...
```
Number:Speed   Weather_Wind_Speed_Mph       "Wind Speed [%.0f mph]"	        <wind>	(gWeather)  {channel="weatherunderground:weather:71ed51c0:current#windSpeed"}
Number:Speed   Weather_Wind_Gust_Mph        "Wind Speed (gust) [%.0f mph]"	<wind>	(gWeather)  {channel="weatherunderground:weather:71ed51c0:current#windGust"}
```

... are getting values and displaying properly...
![image](https://user-images.githubusercontent.com/21967853/37562884-ffd1c598-2a48-11e8-91d2-6c2bf238199c.png)

...but appear to be occasionally causing an error:
```
2018-03-18 00:07:36.535 [ERROR] [clipse.smarthome.core.internal.events.EventHandler] - Creation of ESH-Event failed, because one of the registered event factories has thrown an exception: Error invoking #valueOf(String) on class 'org.eclipse.smarthome.core.library.types.QuantityType' with value '0E-22 mph'.
java.lang.IllegalStateException: Error invoking #valueOf(String) on class 'org.eclipse.smarthome.core.library.types.QuantityType' with value '0E-22 mph'.
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.parseSimpleClassName(ItemEventFactory.java:178) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.parseType(ItemEventFactory.java:146) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.getState(ItemEventFactory.java:124) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.createStateEvent(ItemEventFactory.java:111) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.createEventByType(ItemEventFactory.java:77) ~[?:?]
        at org.eclipse.smarthome.core.events.AbstractEventFactory.createEvent(AbstractEventFactory.java:50) ~[?:?]
        at org.eclipse.smarthome.core.internal.events.EventHandler.createESHEvent(EventHandler.java:121) ~[?:?]
        at org.eclipse.smarthome.core.internal.events.EventHandler.handleEvent(EventHandler.java:95) ~[?:?]
        at org.eclipse.smarthome.core.internal.events.EventHandler.handleEvent(EventHandler.java:72) ~[?:?]
        at org.eclipse.smarthome.core.internal.events.ThreadedEventHandler.lambda$0(ThreadedEventHandler.java:71) ~[?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.GeneratedMethodAccessor321.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.parseSimpleClassName(ItemEventFactory.java:169) ~[?:?]
        ... 10 more
Caused by: java.lang.IllegalArgumentException: E not recognized (in 0 E-22 mph at index 2)
        at tec.uom.se.quantity.Quantities.getQuantity(Quantities.java:80) ~[?:?]
        at org.eclipse.smarthome.core.library.types.QuantityType.<init>(QuantityType.java:78) ~[?:?]
        at org.eclipse.smarthome.core.library.types.QuantityType.valueOf(QuantityType.java:121) ~[?:?]
        at sun.reflect.GeneratedMethodAccessor321.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.parseSimpleClassName(ItemEventFactory.java:169) ~[?:?]
        ... 10 more
```

--- a/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/library/types/QuantityTypeTest.java
+++ b/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/library/types/QuantityTypeTest.java
@@ -70,6 +70,7 @@ public class QuantityTypeTest {
         new QuantityType<>("3 µs");
         new QuantityType<>("3km/h");
         new QuantityType<>("1084 hPa");
+        new QuantityType<>("0E-22 m");
         QuantityType.valueOf("2m");
     }
 
@@ -222,6 +223,18 @@ public class QuantityTypeTest {
     @Test(expected = ArithmeticException.class)
     public void testDivide_Zero() {
         new QuantityType<>("4 m").divide(QuantityType.ZERO);
+    }
+
+    @Test
+    public void testExponentials() {
+        QuantityType<Length> exponential = new QuantityType<>("10E-2 m");
+        assertEquals(exponential, new QuantityType<>("10 cm"));
+
+        exponential = new QuantityType<>("10E+3 m");
+        assertEquals(exponential, new QuantityType<>("10 km"));
+
+        exponential = new QuantityType<>("10E3 m");
+        assertEquals(exponential, new QuantityType<>("10 km"));
     }
 
 }

--- a/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/QuantityType.java
+++ b/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/QuantityType.java
@@ -25,6 +25,7 @@ import javax.measure.Unit;
 import javax.measure.UnitConverter;
 import javax.measure.quantity.Dimensionless;
 
+import org.apache.commons.lang.ArrayUtils;
 import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
 import org.eclipse.smarthome.core.library.unit.SmartHomeUnits;
@@ -74,7 +75,8 @@ public class QuantityType<T extends Quantity<T>> extends Number
         String[] constituents = value.split(UNIT_PATTERN);
 
         // getQuantity needs a space between numeric value and unit
-        String formatted = String.join(" ", constituents);
+        constituents = (String[]) ArrayUtils.add(constituents, constituents.length - 1, " ");
+        String formatted = String.join("", constituents);
         quantity = (Quantity<T>) Quantities.getQuantity(formatted);
     }
 

